File: intrinsic/util/status/get_extended_status.cc
// Copyright 2023 Intrinsic Innovation LLC

#include "intrinsic/util/status/get_extended_status.h"

#include <optional>
#include <string>
#include <utility>

#include "absl/log/log.h"
#include "absl/status/status.h"
#include "absl/strings/cord.h"
#include "google/protobuf/any.pb.h"
#include "google/rpc/code.pb.h"
#include "google/rpc/status.pb.h"
#include "grpcpp/support/status.h"
#include "intrinsic/util/proto/type_url.h"
#include "intrinsic/util/status/extended_status.pb.h"

namespace intrinsic {

std::optional<intrinsic_proto::status::ExtendedStatus> GetExtendedStatus(
    const absl::Status& status) {
  if (status.ok()) return std::nullopt;

  std::optional<absl::Cord> extended_status_payload = status.GetPayload(
      AddTypeUrlPrefix<intrinsic_proto::status::ExtendedStatus>());
  if (!extended_status_payload.has_value()) return std::nullopt;

  intrinsic_proto::status::ExtendedStatus es;
  if (!es.ParseFromCord(*extended_status_payload)) {
    LOG(WARNING) << "Failed to parse ExtendedStatus payload from status: "
                 << status;
    return std::nullopt;
  }

  return es;
}

std::optional<intrinsic_proto::status::ExtendedStatus> GetExtendedStatus(
    const google::rpc::Status& status) {
  if (status.code() == google::rpc::OK) return std::nullopt;

  const std::string extended_status_type_url =
      AddTypeUrlPrefix<intrinsic_proto::status::ExtendedStatus>();

  std::optional<intrinsic_proto::status::ExtendedStatus> extended_status;

  for (const google::protobuf::Any& detail_any : status.details()) {
    if (detail_any.type_url() == extended_status_type_url) {
      intrinsic_proto::status::ExtendedStatus es;
      if (detail_any.UnpackTo(&es)) {
        extended_status = std::move(es);
      }
    }
  }

  return extended_status;
}

std::optional<intrinsic_proto::status::ExtendedStatus> GetExtendedStatus(
    const grpc::Status& status) {
  if (status.ok()) return std::nullopt;

  if (status.error_details().empty()) {
    return std::nullopt;
  }

  google::rpc::Status rpc_status;
  if (!rpc_status.ParseFromString(status.error_details())) {
    return std::nullopt;
  }

  return GetExtendedStatus(rpc_status);
}

}  // namespace intrinsic

File: intrinsic/util/status/get_extended_status.py
# Copyright 2023 Intrinsic Innovation LLC

"""Utility for getting extended status from a gRPC status."""

import grpc
from grpc_status import rpc_status
from intrinsic.util.status import extended_status_pb2


def get_extended_status_from_rpc_error(
    rpc_error: grpc.RpcError,
) -> extended_status_pb2.ExtendedStatus | None:
  """Extracts optional extended status from a gRPC status.

  Typical usage example:

    try:
      my_service_stub.MyCall(request)
    except grpc.RpcError as rpc_error:
      extended_status = get_extended_status_from_rpc_error(rpc_error)
      if extended_status is not None:
        # Do something with the extended status.
      else:
        # Do something with the regular grpc.RpcError.

  Args:
    rpc_error: The gRPC error that might contain an extended status.

  Returns:
    The extended status if rpc_error has one attached, otherwise None.

  Raises:
    RuntimeError: If a rpc status detail contains the ExstendedStatus message
      type but cannot be unpacked.
  """
  status = rpc_status.from_call(rpc_error)
  if status is None:
    return None
  for detail in status.details:
    if detail.Is(extended_status_pb2.ExtendedStatus.DESCRIPTOR):
      es = extended_status_pb2.ExtendedStatus()
      if not detail.Unpack(es):
        raise RuntimeError(
            f"Failed to unpack extended status from rpc status detail: {detail}"
        ) from rpc_error
      return es

  return None
